### DISCUSSION TITLE
Rollback weird behaviour?

### DISCUSSION THREAD

**Discussion by MennoM** [cjmellor/approval#52] [General]
Hiya! I've just been toying around with this plugin and found some weird behaviour with the rollback function, just checking in if this is intended or not.

After making a modification on a model that has the MustBeApproved trait an approval gets created, after approving this change and then making a rollback on the same change, in my opinion, something weird happens

Because the rollback itself is another change on the model, changing the data to it's previous version, another approval gets made about the same interaction:
- the initial approval including the rolled back date, which has the state pending
- a new approval about changed data on the model

Which of these changes needs to be approved now?
Is it possible to not make a modification approval about the reverted data, which is already included in the original rolled-back approval?

**Comment by cjmellor** [comment]
Hi,

Yeah, you're right, it shouldn't do another Approval after doing a Rollback -- makes no sense really.

Perhaps something I over looked and should look into.

Thanks for the heads up.

**Edit**

I've just checked the code and tests and clearly I went into this feature with the idea of approving the rollback... which in hindsight isn't the best DX, so I'll have a look at making a change to where both situations are accounted for.

**Reply by MennoM** [reply]
Thanks for the reply, I also agree that the rollback should be approved, but maybe keep the original approval and only set the rollback date (just my 2 cents).

Also when changing this feature keep in kind the following issue I also found yesterday:

When entities have fields with json data the approval has its new_data json encoded but the original_data not.
This works fine for the original approval, but when rolling back the new_data and original_data fields get swapped, so after approving the roll_back I get an error message "json_decode(): Argument #1 ($json) must be of type string, array given", this model has a casts attribute where its field is cast as json

**Reply by cjmellor** [reply]
Nice one. I'll be sure to look into that too. 

Open to a PR too as can't say exactly when I'll get to it 😬

**Reply by cjmellor** [reply]
@MennoM Any chance you could give me a clearer example of how you got to the error, like list the steps you took?

Thanks

**Reply by MennoM** [reply]
This is done in Laravel 10

Let's say I've got the following Model:
class Post extends Model
{
    use MustBeApproved;

    protected $fillable = [
        "config",
    ];

    protected $casts = [
        "config" => "json",
    ];
}

Now let's assume we're making different API calls where we change this Model and approve the modifications, the creation works fine
`
        $post = new Post();
        $post->config = ['checked' => false];
        $post->save();
`

When updating the post (which needs an approval)
`
        $post = Post::find([id]);
        $post->config = ['checked' => true];
        $post->save();
`

I get the following approvals, which I already approved in this image, but you can already see the discrepancy in data
![post_approved](https://github.com/cjmellor/approval/assets/3264201/ce9e2144-7974-49b0-8752-da4f6ff1505f)

after rolling back this change I get the following approvals
![post_rolled_back](https://github.com/cjmellor/approval/assets/3264201/263ebea3-6caf-42d4-880c-13c4b1e1dd87)

Now after approving the rolled-back approval (which is pending as well) I get the error "message": "json_decode(): Argument #1 ($json) must be of type string, array given", "exception": "TypeError", "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Casts/Json.php"

**Reply by cjmellor** [reply]
Hey @MennoM 

Thanks for this, it'll help!

Just an FYI, v1.4.4 _hopefully_ covers the issue you originally brought up, regarding rollbacks needs an Approval.

Once I find some time I can look at the other bits you brought up -- though I'm definitely open to a PR for it too.

Thanks

**Reply by MennoM** [reply]
Thanks for the heads up and the quick replies!
I've just created https://github.com/cjmellor/approval/issues/54 for you, hope you can look into that

**Comment by cjmellor** [comment]
Closed by #54
